autopilot/choice_test.go

```diff
@@ -13,6 +13,101 @@ var (
 	nID4 = NodeID([33]byte{4})
 )
 
+// TestWeightedChoiceEmptyMap tests that passing in an empty slice of weights
+// returns an error.
+func TestWeightedChoiceEmptyMap(t *testing.T) {
+	t.Parallel()
+
+	var w []float64
+	_, err := weightedChoice(w)
+	if err != ErrNoPositive {
+		t.Fatalf("expected ErrNoPositive when choosing in "+
+			"empty map, instead got %v", err)
+	}
+}
+
+// TestWeightedChoiceSingleIndex tests that choosing randomly in a slice with one
+// positive element always returns that one index.
+func TestWeightedChoiceSingleIndex(t *testing.T) {
+	t.Parallel()
+
+	w := []float64{0, 0.5, 0, 0}
+	for i := 0; i < 1000; i++ {
+		res, err := weightedChoice(w)
+		if err != nil {
+			t.Fatalf("unable to get choice: %v", err)
+		}
+
+		if res != 1 {
+			t.Fatalf("expected %x, got %x", nID1, res)
+		}
+	}
+}
+
+func assertChoice(t *testing.T, w []float64,
+	iterations int, expChoices []int) {
+	t.Helper()
+
+	chosen := make(map[int]int)
+	for i := 0; i < iterations; i++ {
+		res, err := weightedChoice(w)
+		if err != nil {
+			t.Fatalf("unable to get choice: %v", err)
+		}
+		chosen[res]++
+	}
+
+	// Since this is random we check that the number of times chosen is
+	// within %20 of the expected value.
+	sum := 0
+	for i, exp := range expChoices {
+		v := chosen[i]
+		sum += v
+		expHigh := exp + exp/5
+		expLow := exp - exp/5
+		if v < expLow || v > expHigh {
+			t.Fatalf("expected index %d to be chosen ~%d times, was %d",
+				i, exp, v)
+		}
+	}
+
+	// The sum of choices must be exactly iterations of course.
+	if sum != iterations {
+		t.Fatalf("expected a sum of %d, got %d", iterations, sum)
+	}
+
+}
+
+// TestWeightedChoiceDistribution asserts that the weighted choice algorithm
+// chooses among indexes according to their scores.
+func TestWeightedChoiceDistribution(t *testing.T) {
```

Comment:
I think a similar technique as outlined above (using `testing/quick`) can be applied here as well. One way to go about it would be to compare two ideas: one that will always pick option A 25% of the time (as an example) and the other which is actually the `weightedChoice` function. We'd that assert the two to be within some statistical difference of each other.

Reply:
Done!